### PR TITLE
chore(gradle): build manual zip packs

### PR DESCRIPTION
**Pull request by miurahr**
- Add new task `manualPacks` to produce `build/manuals/manual<Lang>.zip`
- Pack manual zips in OmegaT distribution/installation instead of plain files
- Tweak task dependencies to avoid unwanted duplicated run

## Pull request type

- Build and release changes -> [build/release]

## What does this PR change?

- Clean docs/${lang}/*.html, docs/${lang}/images and docs/${lang}/_wh other than instant-start and OmegaT.css
- Update doc_src/build.xml to accept target folder
- Update doc_src/build.xml to add instant-start-pictures to copy images for instant-start

## Other information

Discussed in dev-list
https://sourceforge.net/p/omegat/mailman/omegat-development/thread/xe9mv6M8stttjRgudIj5zEUiGO4UNyIknfxEEJeJIa6eqvTsuXIqt0fge3G5S0avzGovvJcyAKZ7zVyV5s-FxTEG1pWj-qsfFasIeekM3wk%3D%40northside.tokyo/#msg37870550

### CODE REVIEW

**Review by brandelune — COMMENTED** [review]
Sincere apologies. I don't have enough time to check now.

**Comment by miurahr** [comment]
With this changes, installation folder become very clean

```
OmegaT  OmegaT.bat  OmegaT.jar  docs  images  lib  linux-install.sh  modules  omegat.desktop  omegat.kaptn  plugins  scripts
```

docs folder contains
```

be                  cs                el               eu      gl.zip  index.html  ko             pt          readme_be.txt  readme_de.txt  readme_hr.txt  readme_mfe.txt  readme_ru.txt     readme_tr.txt     sc      zh_CN.zip
be.zip              cs.zip            el.zip           fi      hr      it          libraries.txt  pt.zip      readme_ca.txt  readme_es.txt  readme_hu.txt  readme_nl.txt   readme_sh.txt     readme_uk.txt     sv
ca                  cy                en               fi.zip  hr.zip  it.zip      mfe            pt_BR       readme_co.txt  readme_eu.txt  readme_ia.txt  readme_no.txt   readme_sk.txt     readme_zh_CN.txt  tr
ca.zip              de                en.zip           fr      hu      ja          nl             pt_BR.zip   readme_cs.txt  readme_fi.txt  readme_it.txt  readme_pl.txt   readme_sl.txt     readme_zh_TW.txt  uk
changes.txt         de.zip            es               fr.zip  ia      ja.zip      nl.zip         readme.txt  readme_cy.txt  readme_fr.txt  readme_ja.txt  readme_pt.txt   readme_sq.txt     ru                uk.zip
```

Now folder `en` has a page which is shown as first step.
There is also a file `version.properties`
```
OmegaT.css  first_steps.html   version.properties
```

File: build.gradle [hunk 1]
```diff
@@ -410,55 +397,157 @@ task genDocIndex(type: Copy) {
...
+    description = 'Build PDF manuals for all languages. Requires Docker.'
+}
+
+task manualHtmls {
+    description = 'Build HTML manuals for all languages. Requires Docker.'
+}
+
+task manualPacks {
+    description = 'Build HTML manual packs for all languages. Requires Docker.'
+}
```
Comment (miurahr) [inline build.gradle:416]:
```suggestion
```

We can remove the task.

**Comment by miurahr** [comment]
Reviewer should look following files

- build.gradle
    - There is new tasks named `manualZip<Lang>`  which produce zip files
- doc_src/build.xml
    - Update to accept TARGET folder to generate. It is necessary to change a folder where artifacts generated.
    - Same docker image is used to generate  manuals and first-step pages. target argument change a direction.
- ci/azure-pipelines/build-doc-steps.yml  ci/azure-pipelines/publish-manual-snapshot.yml    
    - For convenience of manual writers and l10n members, it provide a snapshot html manuals
- src/org/omegat/help/Help.java  src/org/omegat/util/CheckVersion.java
    - Add method to show zipped manuals on users' web browser
    - Compare manual version and OmegaT version, which ignore patch and update versions.

**Comment by miurahr** [comment]
I'd like to recreate commits history to be best for review.

**Comment by miurahr** [comment]
There is only English version of manual has a version number 6.1.0, so  `Help>manual` show a zipped manual only in English locale.

**Comment by Kazephil** [comment]
> There is only English version of manual has a version number 6.1.0, so `Help>manual` show a zipped manual only in English locale.

The French version also has a 6.1.0 version number.

I'll try to take a closer look at the rest of the PR later today.

**Review by Kazephil — COMMENTED** [review]
I found time to try this out a little, and had the following issues:

- There is no `index.html` file in the `docs` subfolder, so trying to open `index.html` in the `OmegaT` folder results in a "Your file couldn’t be accessed" error.
- The `en` subfolder in the `docs` folder is empty.
- Running OmegaT and clicking Help --> User manual (or pressing F1) still sends me to the https://omegat.sourceforge.io/manual-latest/ URL instead of showing the contents of the `en.zip` file in the `manuals` subfolder.

**Comment by miurahr** [comment]
@Kazephil  how to build?

**Comment by Kazephil** [comment]
> @Kazephil  how to build?

I tried the `InstallDist` and the `DistZip` tasks (from Intellij IDEA).

**Comment by miurahr** [comment]
I've tried to reproduce a situation. I run `gradlew clean` and manually removes `docs` folder.
when running `installDist` tasak, gradle select `firstSteps` tasks at first.  It includes `firstStepsEn` task that produce `docs/greetings/en/first_steps.html`.
then gradle seelct `instantStarts` tasks.  It also produces `docs/greetings/<lang>/first_steps.xml`. 
There may not be `docs/en/` folder.

After generating `first_steps.html`, gradle select `compileJava` and `jar` tasks which provide `libs/OmegaT.jar`. then start `manualHtmls` tasks and `manualZip*` tasks.

You can find `build/docs/manual/index.html` and raw html contents in `build/docs/manual/en/foo.html`.
You can also find `docs/manual/en.zip` zipped manuals.

We can take a way that we leave a contents of greetings in project `docs/<lang>` folder, as same as current master branch,  such as `first_steps.html`, `OmegaT.css` and optional `images/InstantStart_1.png` files and remove dependency of firstSteps and instantStartGuides from build path.

After overall process finished, `installDist` will show folder structure like

```
build/install/installDist/ 
    --   docs/greetings/en/
    --   docs/manuals/en.zip
    --   docs/manuals/ja.zip
    --  docs/readme.txt
    --  OmegaT.jar
    --  lib/ 
    --  OmegaT
    -- plugins/
    --  scripts/
```

For the reason of build performance, I've fixed dependency graph not to re-run if you finished manual generation once, gradle will skip build as "UP-TO-DATE" after second time unless you run `clean`.
Now I also research a way to realize `build cache` for manuals/greetings without commit artifact in source repository (bad habit).  One idea is publish documents and languagetools as a language pack and OmegaT is loading it in runtime.

We can separate build process of OmegaT distribution and document distribution when we realize language pack.

**Comment by miurahr** [comment]
> There is no index.html file in the docs subfolder

This is a change intended. you can open `build/docs/manual/index.html` instead for review process.
It is why there is no way to open zipped content on web browser from index.html.

> The en subfolder in the docs folder is empty.

This is a change intended.  there are two sub folders `greetings` and `manuals`.

> Running OmegaT and clicking Help --> User manual (or pressing F1) still sends me to the

I design to open local contents by extracting zip and show it, so it is a bug.

**Comment by miurahr** [comment]
> > Running OmegaT and clicking Help --> User manual (or pressing F1) still sends me to the
> 
> I design to open local contents by extracting zip and show it, so it is a bug.

I've pushed the fix.

**Comment by miurahr** [comment]
`build/docs/manuals/index.html` is generated by `webManual` task or `updateManuals` task as same as before.

**Comment by miurahr** [comment]
Now changes are completed for test.  Please test with `-Duser.language=fr` or `-Duser.langauge=en` for checking behavior to show zip contents, because these two  manuals are  up-to-date and bundled.

**Comment by Kazephil** [comment]
I tried running `clean` and `installDist` again.

> 
> After overall process finished, `installDist` will show folder structure like
> 
> ```
> build/install/OmegaT/ 
>     --   docs/greetings/en/
>     --   docs/manuals/en.zip
>     --   docs/manuals/ja.zip
>     --  docs/readme.txt
>     --  OmegaT.jar
>     --  lib/ 
>     --  OmegaT
>     -- plugins/
>     --  scripts/
> ```

Yes, I get essentially the same structure (abridged to the essentials):
```
build
└── install
    └── OmegaT
        ├── index.html
        ├── join.html
        ├── OmegaT.jar
        ├── (+ the license files, .desktop file,
        |      and install/launch shell and batch scripts)
        ├── docs
        │   ├── greetings
        │   |   ├── en
        │   |   └── ja
        │   |   (and so on for the other languages)
        │   └── manuals
        │   |   ├── en.zip
        │   |   └── ja.zip
        │   |   (and so on for the other languages, but only zip files)
        │   ├── changes.txt
        │   ├── contributors.txt
        │   ├── libraries.txt
        │   └── readme.txt (+all other language versions)
        ├── images
        ├── lib
        ├── modules
        ├── plugins
        └── scripts
```

> We can separate build process of OmegaT distribution and document distribution when we realize language pack.

So if there are fixes or updates to the documentation, users could download just the language pack and get the updated documentation?

> > There is no index.html file in the docs subfolder
> 
> This is a change intended. you can open `build/docs/manual/index.html` instead for review process. It is why there is no way to open zipped content on web browser from index.html.

As shown in the directory tree above, I don't see a `build/docs/manual/index.html` file in the `manuals` subdirectory. I only see an `index.html` file in the `.zip` files.

Also, the `index.html` file in the `OmegaT` directory points to `docs/index.html`, so either:
- we point it to `docs/manuals/index.html` (if that file is supposed to exist in the `manuals` subdirectory), or
- we remove `index.html` from the `OmegaT` directory completely (if not having an `index.html` file in the `manuals` subdirectory is correct.)

The second option is probably a bad idea, though.

I just looked at the actual `index.html` in the `docs` subdirectory of the standard distribution. It correctly points to the local HTML documentation.

Side note: I'm going to have to do something about the links to the PDFs, though. We don't have PDFs for the current version, and the links for older versions are wrong. (I'll open a separate issue for that.)

> > The en subfolder in the docs folder is empty.
> 
> This is a change intended. there are two sub folders `greetings` and `manuals`.

Indeed. I no longer have the `en` subdirectory in the `docs` directory after cleaning and re-running `installDist`.

> > Running OmegaT and clicking Help --> User manual (or pressing F1) still sends me to the
> 
> I design to open local contents by extracting zip and show it, so it is a bug.

Indeed. It doesn't seem to have been a build issue. I still get the same behavior after running `clean` and `installDist` again.

**Comment by Kazephil** [comment]
> Now changes are completed for test. Please test with `-Duser.language=fr` or `-Duser.langauge=en` for checking behavior to show zip contents, because these two manuals are up-to-date and bundled.

Looks like I typed my reply while you were busy fixing things again.

I'll update my local branch and test your fix as soon as I can.

**Comment by Kazephil** [comment]
> Now changes are completed for test. Please test with `-Duser.language=fr` or `-Duser.langauge=en` for checking behavior to show zip contents, because these two manuals are up-to-date and bundled.

OK, I updated my local version, and tried again.

For the sake of completeness, I tested with `-Duser.language=en`, ` -Duser.language=fr`, as well as with `-Duser.language=ja` to test the behaviour with a language that doesn't thave the 6.1.0 version of the manual.

Results:
- Both `-Duser.language=en`, ` -Duser.language=fr`, correctly unzipped the local manual and displayed it in the browser.
- With `-Duser.language=ja`, I still get the link to the page with the old (4.3 or earlier) versions of the manual.

I'll start working on fixing `index.html` later this evening.

**Comment by Kazephil** [comment]
> > * With `-Duser.language=ja`, I still get the link to the page with the old (4.3 or earlier) versions of the manual.
> 
> This is a correct behavior based on current versioning policy, because master branch is defined as _Beta . Beta is defiled as latest version and jump to latest manual. Current released last latest version is 5.7.1. OmegaT 5.7.1 published without updating Japanese manuals, so it point to 4.3 or earlier manual.

Oh. That's an… err… "interesting" way to approach versioning for the manual.

I would expect the manual to link to the closest equal or lower version **_number_**, regardless of labels such as "standard", "beta", or "latest".

But that's a point best discussed elsewhere.
I understand the reasoning with respect to the proposal in this specific branch. Thanks for the clarification.

**Comment by miurahr** [comment]
Any further review feedback or confirmation?

**Comment by miurahr** [comment]
> > We can separate build process of OmegaT distribution and document distribution when we realize language pack.
> 
> So if there are fixes or updates to the documentation, users could download just the language pack and get the updated documentation?

Yes, exactly. It is what I want to realize.

**Comment by miurahr** [comment]
> > > There is no index.html file in the docs subfolder
> > 
> > 
> > This is a change intended. you can open `build/docs/manual/index.html` instead for review process. It is why there is no way to open zipped content on web browser from index.html.
> 
> As shown in the directory tree above, I don't see a `build/docs/manual/index.html` file in the `manuals` subdirectory. I only see an `index.html` file in the `.zip` files.

'docs/manual/index.html' is generated by "genDocIndex" task which is called by "webManual".
In here, "webManual" is called by "publishManual".

**Comment by miurahr** [comment]
Waiting reviewers to change status as confirmed, or reviewer give up, or passed 1 month without any response for merge.

**Comment by brandelune** [comment]
After a git fetch on master I'm not seeing your branch.

**Comment by miurahr** [comment]
I can reproduce it on my local. I'd like to update it.

```
Build file '…/build.gradle' line: 655
```

**Comment by Kazephil** [comment]
@miurahr 

I just tried your latest fix. The branch now builds correctly.

The `docs` subdirectory looks like this:
```
.
├── doc-license.txt
├── en
│   └── images --> empty directory
├── greetings
│   ├── be
│   │   ├── first_steps.html
│   │   ├── images
│   │   │   ├── InstantGuide_1.png
│   │   │   └── InstantGuide_2.png
│   │   └── OmegaT.css
│   │
│   └── (and so on for each language)
│   
└── manuals
    ├── en.zip
    └── fr.zip
```
 I'm not sure why there's a separate `en` directory with an empty `images` directory, but everything else looks good.

If I run OmegaT with the interface in English or French, calling help loads the latest manual in my default browser from the zip file in the `manuals` subdirectory.

If I run OmegaT with the interface in Japanese, calling help goes to the OmegaT site page with a list of available manuals and their versions.

**Comment by miurahr** [comment]
@Kazephil 

> The `docs` subdirectory looks like this:
> 
> ```
> .
> ├── doc-license.txt
> ├── en
> │   └── images --> empty directory
> ```
> 
> I'm not sure why there's a separate `en` directory with an empty `images` directory, but everything else looks good.

Thank you for the check and feedback.
I think it is because a way to produce folders.

Current script `build.gradle` produce   `/docs/greetings` and `/docs/manuals` then copied into `build/install/OmegaT` when `installDist` task is ordered.

`/docs` does not clear when `clean` task called. so there can be some unwanted folder/data there.

It is better to produce manual zip files and greetings files in `build/docs/manuals` and `build/docs/greetings` then sync into `/docs`.

I'd like to change it.

**Comment by miurahr** [comment]
I think the issue that unwanted folder shown in installation is solved.

**Comment by Kazephil** [comment]
I'm still getting the same directory structure with the empty `docs/en/images` subdirectory.

Here are the steps I took:
1. Delete the build directory completely to make sure there is nothing left over from the previous attempt.
   (It looks like the `installDist` task deletes it before building, but I wanted to make sure there was nothing there.)
2. Gradle `clean` (via IDEA).
3. Gradle `installDist` (also via IDEA)

I also tried the `distZip` task, and the resulting archive also had the `en/images` subdirectory .

**Comment by miurahr** [comment]
Please check `git status` does not report unwanted files.

**Comment by Kazephil** [comment]
> Please check `git status` does not report unwanted files.

The output of `git status`is:

```
On branch topic/miurahr/build/manula-zipped-distribution
Your branch is up to date with 'miurahr/topic/miurahr/build/manula-zipped-distribution'.

nothing to commit, working tree clean

```

**Comment by miurahr** [comment]
I'd like to change to not to produce `docs/` folder at all.  It is only used for test run purpose when `./gradlew run`, so I can change a logic to find `docs` folder under `build/docs`

**Comment by miurahr** [comment]
Now it does not produce/manipulate anything in `docs/` folder.
If you observes somethings in `docs/` it is not related to the PR.

**Pull request merged**